store/src/sqlite/model.rs

```diff
@@ -0,0 +1,1184 @@
+use std::{collections::BTreeSet, marker::PhantomData};
+
+use anyhow::{anyhow, Result};
+use async_trait::async_trait;
+use bytes::Bytes;
+use ceramic_core::{EventId, RangeOpen};
+use cid::Cid;
+use iroh_bitswap::Block;
+use iroh_car::{CarHeader, CarReader, CarWriter};
+use itertools::{process_results, Itertools};
+use multihash::{Code, MultihashDigest};
+use recon::{AssociativeHash, HashCount, InsertResult, Key, ReconItem, Sha256a};
+use sqlx::Row;
+use tracing::instrument;
+
+use crate::{DbTx, SqlitePool};
+
+/// Unified implementation of [`recon::Store`] and [`iroh_bitswap::Store`] that can expose the
+/// individual blocks from the CAR files directly.
+#[derive(Clone, Debug)]
+pub struct ModelStore<H>
+where
+    H: AssociativeHash,
+{
+    pool: SqlitePool,
+    hash: PhantomData<H>,
+}
+
+#[derive(Debug)]
+struct BlockRow {
+    cid: Cid,
+    root: bool,
+    bytes: Vec<u8>,
+}
+
+impl<H> ModelStore<H>
+where
+    H: AssociativeHash + std::convert::From<[u32; 8]>,
+{
+    /// Create an instance of the store initializing any neccessary tables.
+    pub async fn new(pool: SqlitePool) -> Result<Self> {
+        let mut store = ModelStore {
+            pool,
+            hash: PhantomData,
+        };
+        store.create_table_if_not_exists().await?;
+        Ok(store)
+    }
+
+    /// Initialize the recon table.
+    async fn create_table_if_not_exists(&mut self) -> Result<()> {
+        const CREATE_STORE_KEY_TABLE: &str = "CREATE TABLE IF NOT EXISTS model_key (
+            key BLOB, -- network_id sort_value controller StreamID height event_cid
+            ahash_0 INTEGER, -- the ahash is decomposed as [u32; 8]
+            ahash_1 INTEGER,
+            ahash_2 INTEGER,
+            ahash_3 INTEGER,
+            ahash_4 INTEGER,
+            ahash_5 INTEGER,
+            ahash_6 INTEGER,
+            ahash_7 INTEGER,
+            value_retrieved BOOL, -- indicates if we have the value
+            PRIMARY KEY(key)
+        )";
+        const CREATE_VALUE_RETRIEVED_INDEX: &str =
+            "CREATE INDEX IF NOT EXISTS idx_key_value_retrieved
+            ON model_key (key, value_retrieved)";
+
+        const CREATE_MODEL_BLOCK_TABLE: &str = "CREATE TABLE IF NOT EXISTS model_block (
+            cid BLOB, -- the cid of the Block as bytes no 0x00 prefix
+            key BLOB, -- network_id sort_value controller StreamID height event_cid
+            idx INTEGER, -- the index of the block in the CAR file
+            root BOOL, -- when true the block is a root in the CAR file
+            bytes BLOB, -- the Block
+            PRIMARY KEY(cid, key)
```

Comment:
Does the order of the columns in the key matter? We want all the cids for key contiguous for a key if possible. Would `PRIMARY KEY(key, cid) ` do that?

Reply:
agree that this makes sense. I changed it and it was interesting (as described below). It could definitely change the disk layout and it should keep all the blocks for a key closer together which is good.